dali/python/nvidia/dali/pipeline.py

```diff
@@ -1455,7 +1455,82 @@ def _discriminate_args(func, **func_kwargs):
     return ctor_args, fn_args
 
 
-def pipeline_def(fn=None, **pipeline_kwargs):
+def _regroup_args(func, pipeline_def_kwargs, fn_call_kwargs):
+    """Regroup arguments that are directed into Pipeline object construction and those that
+    are passed into pipeline definition function.
+
+    Parameters
+    ----------
+    func : Callable
+        The pipeline definition function that is decorated.
+    pipeline_def_kwargs : Dict
+        Kwargs passed to the @pipeline_def
+    fn_call_kwargs : Dict
+        Kwargs passed when invoking the decorated function
+
+    Returns
+    -------
+    (Dict, Dict)
+        Pipeline kwargs, Function kwargs
+    """
+    ctor_args, fn_kwargs = _discriminate_args(func, **fn_call_kwargs)
+    pipeline_kwargs = {**pipeline_def_kwargs, **ctor_args}  # Merge and overwrite dict
+    return pipeline_kwargs, fn_kwargs
+
+
+def _preprocess_pipe_func(func, conditionals_on):
+    """Transform the pipeline defintion function if the conditionals are enabled
+    """
+    if conditionals_on:
+        return _conditionals._autograph.to_graph(func)
+    else:
+        return func
+
+
+def _preprocess_pipe_object(pipe, conditionals_on, args, fn_kwargs):
+    """Based on the conditional mode status, preprocess the pipeline object before the graph
+    is created.
+    """
+    if conditionals_on:
+        Pipeline.push_current(pipe)
+        pipe._conditionals_enabled = True
+        pipe._condition_stack = _conditionals._ConditionStack()
+        # Add all parameters to the pipeline as "know" nodes in the top scope.
+        for arg in args:
+            if isinstance(arg, DataNode):
+                _conditionals.register_data_nodes(arg)
+        for _, arg in fn_kwargs.items():
+            if isinstance(arg, DataNode):
+                _conditionals.register_data_nodes(arg)
+        Pipeline.pop_current()
+
+
+def _generate_graph(pipe, func, fn_args, fn_kwargs):
+    """Build the graph provided by pipeline definition in `func` within the `pipe`.
+
+    Parameters
+    ----------
+    pipe : Pipeline
+        Target pipeline object
+    func : Callable
+        The pipeline definition that is decorated
+    fn_args : List
+        Positional arguments to `func`
+    fn_kwargs : Dict
+        Kwargs to `func`
+    """
+    with pipe:
+        pipe_outputs = func(*fn_args, **fn_kwargs)
+        if isinstance(pipe_outputs, tuple):
+            po = pipe_outputs
+        elif pipe_outputs is None:
+            po = ()
+        else:
+            po = (pipe_outputs, )
+        pipe.set_outputs(*po)
+
+
+def pipeline_def(fn=None, *, enable_conditionals=False, **pipeline_kwargs):
```

Comment:
As far as I am aware, it should just work based on how keyword only arguments and `**kwargs` work.
https://peps.python.org/pep-3102/
```
For each keyword argument:
    If there is a parameter with the same name as the keyword, then the argument value is assigned to that parameter slot. However, if the parameter slot is already filled, then that is an error.
    Otherwise, if there is a ‘keyword dictionary’ argument, the argument is added to the dictionary using the keyword name as the dictionary key, unless there is already an entry with that key, in which case it is an error.
    Otherwise, if there is no keyword dictionary, and no matching named parameter, then it is an error.
```

I can still double check if you think it is needed.

Reply:
I added a test for no-parenthesis option of decorator.